lib_c/dates_calc.c

```diff
@@ -0,0 +1,340 @@
+/* This file is part of the Dates_calc library. Copyright (C) 2024 Inria,
+   contributors: Louis Gesbert <louis.gesbert@inria.fr>, Raphaël Monat
+   <raphael.monat@inria.fr>
+
+   Licensed under the Apache License, Version 2.0 (the "License"); you may not
+   use this file except in compliance with the License. You may obtain a copy of
+   the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+   License for the specific language governing permissions and limitations under
+   the License. */
+
+#include <assert.h>
+#include <stdio.h>
+
+#define BOOL int
+#define FALSE 0
+#define TRUE 1
+
+/* Layout of this C version:
+   - types and functions in this module are prefixed with [dc_]
+   - dates and periods are manipulated as pointers to the defined structs
+   - functions return through a first [ret] pointer argument, the other
+     arguments are [const].
+   - functions that can fail return the [dc_success] type. What is stored into
+     [ret] is unspecified when that is [dc_error].
+   - it is expected that [ret] and other arguments may overlap, so, in the code
+     below, a field in [ret] should never be written before the same field in
+     any other argument of the same type is read.
+*/
+
+typedef enum dc_success {
+  dc_error, dc_ok
+} dc_success;
+
+typedef enum dc_date_rounding {
+  dc_date_round_up,
+  dc_date_round_down,
+  dc_date_round_abort
+} dc_date_rounding;
+
+
+typedef struct dc_date {
+  long int year;
+  unsigned long int month;
+  unsigned long int day;
```

Comment:
Nitpicking: I think both months and days can be stored as `unsigned char`?

Reply:
Hm, indeed, but I'll need to double-check that no intermediate calculations can overflow. Also, it might be less efficient on modern architectures.

Reply:
From a reread of the code, it should be ok (we always use a larger type for the results of arithmetic on them).
I still feel safer keeping the same type rather than multiplying different integer bitlengths, though...

Reply:
No worries!